### PR TITLE
[C++] [lua] Add equipment slot tracking to enchantment item usage, audit ring enchantments

### PR DESCRIPTION
**Pull request by Skold177**
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

Sets up onItemUse and onItemUnequip to pass the equipment slot so that enchantments in the ring/earring slots stack and remove their effects properly. Also closes an edge case where a subType in the item effect could unintentionally be fed as a itemID (which is part of the reason this didn't work correctly to begin with)

Test included.

## Steps to test these changes

!additem Armored Ring x2
/checkparam yourself
note defense
equip them both, see items stack
remove one, see one enchantment leave + lose 9 defense (1+ 8 from the enchantment)
remove the other

### CODE REVIEW

**Comment by Skold177** [comment]
Don't 100% know if this is the best way to do it this, so opening as a draft for feedback, if we like it, i'll undraft!

File: src/map/status_effect_container.cpp [hunk 1]
```diff
@@ -1524,7 +1524,7 @@ auto CStatusEffectContainer::SetEffectParams(CStatusEffect* StatusEffect) -> voi
 
     // Is an effect from a usable item not caught above.
     // Known use cases: Enchantments without an effect source.
-    else
+    else if (name.empty())
```
Comment (WinterSolstice8) [inline src/map/status_effect_container.cpp:1527]: this doesn't look right, are you sure it's not supposed to be `!name.empty()`

I dont think any effects do have empty names?

Comment (Skold177) [reply src/map/status_effect_container.cpp:1527]: > this doesn't look right, are you sure it's not supposed to be `!name.empty()`
> 
> I dont think any effects do have empty names?

A few in the code base still do - like Pelican Ring

It doesn't provide a sourceType or a sourceTypeParam so its "name" is empty, then it takes the subPower and uses it to look up an itemID - and its name gets populated from that. It's legacy code we could delete if we converted any items missing sourceType over

Comment (WinterSolstice8) [reply src/map/status_effect_container.cpp:1527]: Oh I see, yeah that sounds right then
